compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,456 @@
+            typeid.push_str(&s);
+        }
+
+        // User-defined types
+        ty::Adt(adt_def, substs) => {
+            let mut s = String::new();
+            let def_id = adt_def.0.did;
+            let name = format!("{}", tcx.opt_item_name(def_id).unwrap());
+            let crate_name = format!("{}", tcx.crate_name(def_id.krate));
+            // FIXME(rcvalle): Identify C char and integer type uses and encode them with their
+            // respective builtin type encoding as specified by the Itanium C++ ABI if
+            // encode_c_types is true.
+            if name == "c_void"
+                && (crate_name == "core" || crate_name == "std" || crate_name == "libc")
+            {
+                // C char, integer, and void types
+                s.push('v');
+            } else if adt_def.repr().c() {
+                // <length><name>
+                s.push_str(&format!("{}", name.len()));
+                s.push_str(&name);
+                compress(dict, (ty, TyQ::None), &mut s);
+            } else if adt_def.is_struct() && adt_def.repr().transparent() {
+                // <element-type1>
+                let mut tys = adt_def.all_fields().map(|field| field.ty(tcx, substs));
+                let ty0 = tys.next().unwrap();
+                s.push_str(&encode_ty(tcx, ty0, dict, encode_c_types));
+            } else {
+                // u<length><type-name>[I<element-type1..element-typeN>E] as vendor extended type
+
+                // Start and namespace tags
+                let mut def_path = tcx.def_path(def_id);
+                def_path.data.reverse();
+                for disambiguated_data in &def_path.data {
+                    s.push_str(match disambiguated_data.data {
+                        hir::definitions::DefPathData::TypeNs(_) => "Nt",
+                        hir::definitions::DefPathData::ValueNs(_) => "Nv",
+                        _ => {
+                            bug!("encode_ty: unexpected `{:?}`", disambiguated_data.data);
+                        },
+                    });
+                }
+
+                // Crate disambiguator and name
+                s.push('C');
+                s.push_str(&to_disambiguator(tcx.stable_crate_id(def_path.krate).to_u64()));
+                let crate_name = format!("{}", tcx.crate_name(def_path.krate));
+                s.push_str(&format!("{}", crate_name.len()));
+                s.push_str(&crate_name);
+
+                // Disambiguators and names
+                def_path.data.reverse();
+                for disambiguated_data in &def_path.data {
+                    let num = disambiguated_data.disambiguator as u64;
+                    if num > 0 {
+                        s.push_str(&to_disambiguator(num));
+                    }
+                    let name = format!("{}", disambiguated_data.data);
+                    s.push_str(&format!("{}", name.len()));
+                    s.push_str(&name);
```

Comment:
The Itanium C++ ABI does not specify a mangling for identifiers containing characters outside of _A-Za-z0-9, but Rust supports non-ASCII identifiers/names. Since mangling for non-ASCII characters are unspecified by the Itanium C++ ABI, should I just use them as is or encode them using something like Base64 or Punycode?